runtime/src/chpl-qsbr.c

```diff
@@ -0,0 +1,481 @@
+    return NULL;
+  }
+
+  // If prev is NULL then we have the actual head of the list
+  if (prev == NULL) {
+    node->deferList = NULL;
+    return dnode;
+  }
+
+  // If neither are NULL, we need to correct list so previous points to NULL
+  prev->next = NULL;
+  return dnode;
+}
+
+// Requires spinlock.
+static inline struct defer_node *pop_all_defer_list(struct tls_node *node);
+static inline struct defer_node *pop_all_defer_list(struct tls_node *node) {
+  struct defer_node *head = node->deferList;
+  node->deferList = NULL;
+  return head;
+}
+
+// Requires spinlock.
+static inline void push_defer_list(struct tls_node *node, struct defer_node *dnode);
+static inline void push_defer_list(struct tls_node *node, struct defer_node *dnode) {
+  dnode->next = node->deferList;
+  node->deferList = dnode;
+}
+
+// Requires spinlock.
+static inline struct defer_node *get_recycle_list(struct tls_node *node);
+static inline struct defer_node *get_recycle_list(struct tls_node *node) {
+  return node->recycleList;
+}
+
+// Requires spinlock.
+static inline struct defer_node *pop_recycle_list(struct tls_node *node);
+static inline struct defer_node *pop_recycle_list(struct tls_node *node) {
+  struct defer_node *dnode = node->recycleList;
+  
+  // If empty, return fresh allocation
+  if (dnode == NULL) {
+    return chpl_mem_calloc(1, sizeof(struct defer_node), 0, 0, 0);
+  }
+
+  node->recycleList = dnode->next;
+  return dnode;
+}
+
+static inline void push_recycle_list(struct tls_node *node, struct defer_node *dnode);
+static inline void push_recycle_list(struct tls_node *node, struct defer_node *dnode) {
+  dnode->next = node->recycleList;
+  node->recycleList = dnode;
+}
+
+static inline void delete_data(struct defer_node *dnode);
+static inline void delete_data(struct defer_node *dnode) {
+  if (dnode->numData) {
+      void **arr = dnode->data;
+      for (int i = 0; i < dnode->numData; i++) {
+          chpl_mem_free(arr[i], 0, 0);
+      }
+  }
+  chpl_mem_free(dnode->data, 0, 0);
+}
+
+// Initializes TLS; should only need to be called once.
+static void init_tls(void);
+static void init_tls(void) {
+  struct tls_node *node = chpl_calloc(1, sizeof(struct tls_node));
+  node->epoch = get_global_epoch();
+
+  // Append to head of list
+  struct tls_node *old_head;
+  do {
+    old_head = chpl_qsbr_tls_list;
+    node->next = old_head;
+  } while (!atomic_compare_exchange_weak_uintptr_t(
+      (uintptr_t *) &chpl_qsbr_tls_list, (uintptr_t) old_head, (uintptr_t) node)
+    );
+
+  CHPL_TLS_SET(chpl_qsbr_tls, node);
+}
+
+// Obtains the minimum epoch of all threads.
+static uint64_t safe_epoch(void);
+static uint64_t safe_epoch(void) {
+  uint64_t min = (uint64_t) -1;
+  // Check all remaining threads for whether they have updated to a more recent epoch.
+  for (struct tls_node *node = get_tls_list(); node != NULL; node = node->next) {
+    if (get_tasks(node) > 0) {
+      uint64_t epoch = get_epoch(node);
```

Comment:
It seems obviously OK if we were to get ntasks and epoch in one atomic operation. Can you explain a little bit why it's OK to always get the number of tasks first? I.e. the number of tasks could be 0, and then 1, or vice versa, by the time we get_epoch. Why is that OK?

Reply:
It is under the assumption that threads are not discarded but pooled and that only tasks access protected data. Under this assumption, if a thread has 0 tasks then it need not be taken into account as it will never access the data anyway (in fact, it is considered as 'inactive'), and if it becomes active after our check then it will not see the old data about to be deleted because it at this point the old data is no longer visible. If the thread loses a task after our check, then it won't matter since we are looking for the minimum and we don't give it any special treatment (I.E if it had a lower epoch, then its fine if the minimum drops lower than it would have to, we just end up reclaiming less data this time around)

Reply:
It seems it would be a problem if an activation record (i.e. a task that hasn't been assigned a thread yet) stored the protected data. I think this whole mechanism is for relatively short-lived references anyway, so that's probably OK.

I think your same argument applies even if new threads can be added partway through. The new thread would only see current data, right? Or am I missing something?

Reply:
If you have references to an object from multiple places, unfortunately it will be considered unsafe if marked for deletion. Also yes it does applies to newer threads. A task has to be running to be able access the data so it won't matter if it hasn't been assigned a thread yet.

To put it another way, if any task or thread comes into existence after data has been logically deleted from the list then they are unable to access the logically deleted item and are not at risk when the item is physically deleted